src/back/hlsl/keywords.rs

```diff
@@ -171,4 +174,730 @@ pub const RESERVED: &[&str] = &[
+    "DispatchRaysIndex",
+    "DispatchRaysDimensions",
+    "WorldRayOrigin",
+    "WorldRayDirection",
+    "ObjectRayOrigin",
+    "ObjectRayDirection",
+    "RayTMin",
+    "RayTCurrent",
+    "PrimitiveIndex",
+    "InstanceID",
+    "InstanceIndex",
+    "GeometryIndex",
+    "HitKind",
+    "RayFlags",
+    "ObjectToWorld",
+    "WorldToObject",
+    "ObjectToWorld3x4",
+    "WorldToObject3x4",
+    "ObjectToWorld4x3",
+    "WorldToObject4x3",
+    "dot4add_u8packed",
+    "dot4add_i8packed",
+    "dot2add",
+    "unpack_s8s16",
+    "unpack_u8u16",
+    "unpack_s8s32",
+    "unpack_u8u32",
+    "pack_s8",
+    "pack_u8",
+    "pack_clamp_s8",
+    "pack_clamp_u8",
+    "SetMeshOutputCounts",
+    "DispatchMesh",
+    "IsHelperLane",
+    "AllocateRayQuery",
+    "CreateResourceFromHeap",
+    "and",
+    "or",
+    "select",
+    // DXC resource and other types, from https://github.com/microsoft/DirectXShaderCompiler/blob/18c9e114f9c314f93e68fbc72ce207d4ed2e65ae/tools/clang/lib/AST/HlslTypes.cpp#L441-#L572
+    "InputPatch",
+    "OutputPatch",
+    "PointStream",
+    "LineStream",
+    "TriangleStream",
+    "Texture1D",
+    "RWTexture1D",
+    "Texture2D",
+    "RWTexture2D",
+    "Texture2DMS",
+    "RWTexture2DMS",
+    "Texture3D",
+    "RWTexture3D",
+    "TextureCube",
+    "RWTextureCube",
+    "Texture1DArray",
+    "RWTexture1DArray",
+    "Texture2DArray",
+    "RWTexture2DArray",
+    "Texture2DMSArray",
+    "RWTexture2DMSArray",
+    "TextureCubeArray",
+    "RWTextureCubeArray",
+    "FeedbackTexture2D",
+    "FeedbackTexture2DArray",
+    "RasterizerOrderedTexture1D",
+    "RasterizerOrderedTexture2D",
+    "RasterizerOrderedTexture3D",
+    "RasterizerOrderedTexture1DArray",
+    "RasterizerOrderedTexture2DArray",
+    "RasterizerOrderedBuffer",
+    "RasterizerOrderedByteAddressBuffer",
+    "RasterizerOrderedStructuredBuffer",
+    "ByteAddressBuffer",
+    "RWByteAddressBuffer",
+    "StructuredBuffer",
+    "RWStructuredBuffer",
+    "AppendStructuredBuffer",
+    "ConsumeStructuredBuffer",
+    "Buffer",
+    "RWBuffer",
+    "SamplerState",
+    "SamplerComparisonState",
+    "ConstantBuffer",
+    "TextureBuffer",
+    "RaytracingAccelerationStructure",
+    // DXC templated types, from https://github.com/microsoft/DirectXShaderCompiler/blob/18c9e114f9c314f93e68fbc72ce207d4ed2e65ae/tools/clang/lib/AST/ASTContextHLSL.cpp
+    // look for `BuiltinTypeDeclBuilder`
+    "matrix",
+    "vector",
+    "TextureBuffer",
+    "ConstantBuffer",
+    "RayQuery",
 ];
+
+// DXC scalar types, from https://github.com/microsoft/DirectXShaderCompiler/blob/18c9e114f9c314f93e68fbc72ce207d4ed2e65ae/tools/clang/lib/AST/ASTContextHLSL.cpp#L48-L254
+// + vector and matrix shorthands
+pub const TYPES: &[&str] = &{
+    const L: usize = 23 * (1 + 4 + 4 * 4);
+    let mut res = [""; L];
+    let mut c = 0;
+
+    /// For each scalar type, it will additionally generate vector and matrix shorthands
+    macro_rules! generate {
```

Comment:
Would it be simpler to do this at run time in Rust code? If so, please consider using the `once_cell` crate and simplifying the code.

(It's okay to take on `once_cell` as a dependency, since it's been added to the standard library as of Rust 1.70, so we'll be able to take it out then.)